ext/random/engine_mt19937.c

```diff
@@ -0,0 +1,396 @@
+	s1 ^= (s1 >> 11);
+	s1 ^= (s1 << 7) & 0x9d2c5680U;
+	s1 ^= (s1 << 15) & 0xefc60000U;
+
+	return (uint64_t) (s1 ^ (s1 >> 18));
+}
+
+static zend_long range(php_random_status *status, zend_long min, zend_long max)
+{
+	php_random_status_state_mt19937 *s = status->state;
+
+	if (s->mode == MT_RAND_MT19937) {
+		return php_random_range(&php_random_algo_mt19937, status, min, max);
+	}
+
+	uint64_t r = php_random_algo_mt19937.generate(status) >> 1;
+	/* Legacy mode deliberately not inside php_mt_rand_range()
+	 * to prevent other functions being affected */
+	RAND_RANGE_BADSCALING(r, min, max, PHP_MT_RAND_MAX);
+	return (zend_long) r;
+}
+
+static bool serialize(php_random_status *status, HashTable *data)
+{
+	php_random_status_state_mt19937 *s = status->state;
+	zval t;
+
+	for (uint32_t i = 0; i < MT_N; i++) {
+		ZVAL_STR(&t, php_random_bin2hex_le(&s->state[i], sizeof(uint32_t)));
+		zend_hash_next_index_insert(data, &t);
+	}
+	ZVAL_LONG(&t, s->count);
+	zend_hash_next_index_insert(data, &t);
+	ZVAL_LONG(&t, s->mode);
+	zend_hash_next_index_insert(data, &t);
+	
+	return true;
+}
+
+static bool unserialize(php_random_status *status, HashTable *data)
+{
+	php_random_status_state_mt19937 *s = status->state;
+	zval *t;
+
+	for (uint32_t i = 0; i < MT_N; i++) {
+		t = zend_hash_index_find(data, i);
+		if (!t || Z_TYPE_P(t) != IS_STRING || Z_STRLEN_P(t) != (2 * sizeof(uint32_t))) {
+			return false;
+		}
+		if (!php_random_hex2bin_le(Z_STR_P(t), &s->state[i])) {
+			return false;
+		}
+	}
+	t = zend_hash_index_find(data, MT_N); 
+	if (!t || Z_TYPE_P(t) != IS_LONG) {
+		return false;
+	}
+	s->count = Z_LVAL_P(t);
+	if (s->count > MT_N) {
+		return false;
+	}
+
+	t = zend_hash_index_find(data, MT_N + 1);
+	if (!t || Z_TYPE_P(t) != IS_LONG) {
+		return false;
+	}
+	s->mode = Z_LVAL_P(t);
+	if (s->mode != MT_RAND_MT19937 && s->mode != MT_RAND_PHP) {
+		return false;
+	}
+
+	return true;
+}
+
+const php_random_algo php_random_algo_mt19937 = {
+	sizeof(uint32_t),
+	sizeof(php_random_status_state_mt19937),
+	seed,
+	generate,
+	range,
+	serialize,
+	unserialize
+};
+
+/* {{{ php_random_mt19937_seed_default */
+PHPAPI void php_random_mt19937_seed_default(php_random_status_state_mt19937 *state)
+{
+	zend_long seed = 0;
+
+	if (php_random_bytes_silent(&seed, sizeof(zend_long)) == FAILURE) {
+		seed = GENERATE_SEED();
+	}
+
+	mt19937_seed_state(state, (uint64_t) seed);
+}
+/* }}} */
+
+/* {{{ Random\Engine\Mt19937::__construct() */
+PHP_METHOD(Random_Engine_Mt19937, __construct)
+{
+	php_random_engine *engine = Z_RANDOM_ENGINE_P(ZEND_THIS);
+	php_random_status_state_mt19937 *state = engine->status->state;
+	zend_long seed, mode = MT_RAND_MT19937;
+	bool seed_is_null = true;
+
+	ZEND_PARSE_PARAMETERS_START(0, 2)
+		Z_PARAM_OPTIONAL;
+		Z_PARAM_LONG_OR_NULL(seed, seed_is_null);
+		Z_PARAM_LONG(mode);
+	ZEND_PARSE_PARAMETERS_END();
+
+	switch (mode) {
+		case MT_RAND_PHP:
+			state->mode = MT_RAND_PHP;
+			break;
+		default:
+			state->mode = MT_RAND_MT19937;
+	}
```

Comment:
Should the `mode` parameter also be validated more strictly?

Reply:
fixed: 279227cba04174ba138322b6c6c01f8143ff5ab7